core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -288,6 +288,7 @@ public synchronized IndexShard createShard(int sShardId, ShardRouting routing) {
             indicesLifecycle.afterIndexShardCreated(indexShard);
             settingsService.addListener(indexShard);
             shards = newMapBuilder(shards).put(shardId.id(), indexShard).immutableMap();
+            indexServicesProvider.getIndexingMemoryController().forceCheck();
```

Comment:
I was thinking about it and I think we should start the shard as inactive (it is :)) and only mark it as active + the claiming of memory in IndexShard#skipTranslogRecovery / IndexShard#performTranslogRecovery when we actually open the engine for indexing. We also make sure call IndexingMemoryController.forceCheck before we create the engine, so we always have an up to date buffer size ... 

Re concerns about n^2 problem, it is there but since we limit the amount of concurrent recoveries a node can do, it means it will be done during a longer process. Also because of the limit it is hard to do anything around batching. All in all, I think we're OK.